pkg/controller/service/service_controller.go

```diff
@@ -135,15 +137,28 @@ func New(
 
 	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
 		cache.ResourceEventHandlerFuncs{
-			AddFunc: s.enqueueService,
+			AddFunc: func(cur interface{}) {
+				svc, ok := cur.(*v1.Service)
+				if ok && (wantsLoadBalancer(svc) || needsCleanup(svc)) {
+					s.enqueueService(cur)
+				}
+			},
 			UpdateFunc: func(old, cur interface{}) {
 				oldSvc, ok1 := old.(*v1.Service)
 				curSvc, ok2 := cur.(*v1.Service)
-				if ok1 && ok2 && s.needsUpdate(oldSvc, curSvc) {
+				if ok1 && ok2 && (s.needsUpdate(oldSvc, curSvc) || needsCleanup(curSvc)) {
 					s.enqueueService(cur)
 				}
 			},
-			DeleteFunc: s.enqueueService,
+			DeleteFunc: func(old interface{}) {
+				if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.ServiceLoadBalancerFinalizer) {
+					// No need to handle deletion event if finalizer feature gate is
+					// enabled. Because the deletion would be handled by the update
+					// path when the deletion timestamp is added.
+					return
+				}
+				s.enqueueService(old)
+			},
```

Comment:
The updated handler essentially leads to this workflow:

* AddFunc/UpdateFunc
	* Ensure service is in cache, then
		* Create/resync service with type=LoadBalancer OR Load balancer needs essential update
			* (If enable) Ensure finalizer is added
			* Create/update load balancer if needed
			* Update service status if needed
		* Update/resync service that is deleted with finalizer OR Service changed to non-LB type
			* Delete load balancer
			* (If enable) Ensure finalizer is removed
			* Update service status
			* Remove service from cache
* DeleteFunc
	* Check if service is in cache
		* Service that is deleted without finalizer
			* Delete load balancer
			* Remove service from cache